### PR TITLE
[3D] Allow user adjustment of terrain's elevation

### PR DESCRIPTION
**Pull request by NEDJIMAbelgacem**
## Description

This allows the user to adjust the terrain elevation (moving the terrain up or down)
custom-elevatio-offset
By setting a negative value on terrain elevation offset spinbox, the terrain is moved down. For positive values, it is moved up.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Looking at the code, this applies to ALL terrain types, right?

**Comment by NEDJIMAbelgacem** [comment]
> Looking at the code, this applies to ALL terrain types, right?

@nyalldawson 
Yes, but I only tried it with flat and DEM terrains.

**Comment by nyalldawson** [comment]
Ok, just wasnt sure based on the PR title

File: src/3d/qgscameracontroller.cpp [hunk 1]
```diff
@@ -181,7 +181,10 @@ void QgsCameraController::resetView( float distance )
 void QgsCameraController::setViewFromTop( float worldX, float worldY, float distance, float yaw )
 {
   QgsCameraPose camPose;
-  camPose.setCenterPoint( QgsVector3D( worldX, 0, worldY ) );
+  if ( mTerrainEntity != nullptr )
```
Comment (nyalldawson) [inline src/3d/qgscameracontroller.cpp:184]:
```suggestion
  if ( mTerrainEntity )
```

File: src/app/3d/qgs3dmapconfigwidget.cpp [hunk 2]
```diff
@@ -75,6 +75,7 @@ Qgs3DMapConfigWidget::Qgs3DMapConfigWidget( Qgs3DMapSettings *map, QgsMapCanvas
   spinMapResolution->setClearValue( 512 );
   spinScreenError->setClearValue( 3 );
   spinGroundError->setClearValue( 1 );
+  terrainElevationOffsetSpinBox->setClearValue( 0.0f );
```
Comment (nyalldawson) [inline src/app/3d/qgs3dmapconfigwidget.cpp:78]:
```suggestion
  terrainElevationOffsetSpinBox->setClearValue( 0.0 );
```

(The widget uses double values, not float)

File: src/3d/qgscameracontroller.cpp [hunk 3]
```diff
@@ -275,6 +278,12 @@ void QgsCameraController::updateCameraFromPose( bool centerPointChanged )
       mCameraPose.setCenterPoint( QgsVector3D( intersectionPoint ) );
       mCameraPose.updateCamera( mCamera );
     }
+    else
+    {
+      QgsVector3D centerPoint = mCameraPose.centerPoint();
+      centerPoint.set( centerPoint.x(), mTerrainEntity->terrainElevationOffset(), centerPoint.z() );
```
Comment (wonder-sk) [inline src/3d/qgscameracontroller.cpp:284]: mTerrainEntity may be null here...

File: src/3d/qgs3dmapsettings.cpp [hunk 4]
```diff
@@ -411,12 +415,12 @@ void Qgs3DMapSettings::resolveReferences( const QgsProject &project )
 
 QgsVector3D Qgs3DMapSettings::mapToWorldCoordinates( const QgsVector3D &mapCoords ) const
 {
-  return Qgs3DUtils::mapToWorldCoordinates( mapCoords, mOrigin );
+  return Qgs3DUtils::mapToWorldCoordinates( mapCoords, mOrigin ) + QgsVector3D( 0.0f, mTerrainElevationOffset, 0.0f );
```
Comment (wonder-sk) [inline src/3d/qgs3dmapsettings.cpp:418]: Hmm I am not sure if this is correct... I think map-world coordinates should still have the conversion unaffected by the vertical offset for terrain...

**Review by wonder-sk — COMMENTED** [review]
I thought this would be an option only for the flat terrain :-) Not sure how others feel about it...

Does the altitude clamping work correctly if the flat terrain has non-zero offset? You should probably implement heightAt() virtual function in QgsFlatTerrainGenerator...?

File: src/3d/qgs3dmapsettings.cpp (hunk 4 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgs3dmapsettings.cpp:418]: That depends on whether we want the world coordinates in this context to be the real world coordinates in the opengl scene or disregarding the terrain elevation. I think the elevation offset should be included Qgs3DUtils::mapToWorldCoordinates but I'm not sure how to do that

**Comment by nyalldawson** [comment]
> I thought this would be an option only for the flat terrain :-) Not sure how others feel about it...

Given that it already works for the other types, I don't see any reason to restrict it in the UI to flat terrain only. (E.g. I can see that it would be useful in some cases to manually lower a DEM terrain by a couple of points in order to avoid the ugly visual glitches you get when both the terrain and an object are sitting at exactly the same height.)

**Comment by nyalldawson** [comment]
@NEDJIMAbelgacem can you add some tests too please?

File: src/3d/qgscameracontroller.cpp (hunk 3 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgscameracontroller.cpp:284]: There is an if block that checks if mTerrainEntity is null that contains this else block

**Comment by NEDJIMAbelgacem** [comment]
> Does the altitude clamping work correctly if the flat terrain has non-zero offset? You should probably implement heightAt() virtual function in QgsFlatTerrainGenerator...?

Altitude clamping currently doesn't work. I will try to figure out if that can be implemented without troubles

File: src/3d/qgs3dmapsettings.cpp (hunk 4 above)
Comment (NEDJIMAbelgacem) [reply src/3d/qgs3dmapsettings.cpp]: After some thoughts, I think you are right. The terrain elevation offset should only influence the terrain not apply a translation on the world orign. I reverted baack these modifications

**Comment by NEDJIMAbelgacem** [comment]
> You should probably implement heightAt() virtual function in QgsFlatTerrainGenerator...?
According to the heightAt doc it should just return the height in CRS which is different than the world coordinates
`    //! Returns height at (x,y) in terrain's CRS
    virtual float heightAt( double x, double y, const Qgs3DMapSettings &map ) const;`

**Comment by NEDJIMAbelgacem** [comment]
Hi @nyalldawson 
I discussed things with @wonder-sk . It seems that you can already fix the ugly visual glitches by changing the height from the object's symbol in the vector layer settings. So there is no need to duplicate that behaviour with the terrain elevation offset.
Now I need to add altitude clamping into account, I'm going to implement that before merging.

**Comment by NEDJIMAbelgacem** [comment]
Hi @nyalldawson 
I fixed the issue you mentioned on the eye dome lighting PR  (#39760). It was an issue with the precision of the depth map.
You can now move the flat terrain upo and navigate more easily with the camera

File: src/3d/qgscameracontroller.cpp (hunk 3 above)
Comment (wonder-sk) [reply src/3d/qgscameracontroller.cpp:284]: Ah right... sorry

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**